parlai/tasks/dialogue_nli/agents.py

```diff
@@ -0,0 +1,100 @@
+#!/usr/bin/env python3
+
+# Copyright (c) Facebook, Inc. and its affiliates.
+# This source code is licensed under the MIT license found in the
+# LICENSE file in the root directory of this source tree.
+"""Dialogue safety related datasets and teachers."""
+
+import copy
+import json
+import os
+import random
+import sys as _sys
+
+import parlai.core.build_data as build_data
+from parlai.core.message import Message
+from parlai.core.teachers import FixedDialogTeacher
+
+from .build import build
+
+
+class DialogueNliTeacher(FixedDialogTeacher):
+    def __init__(self, opt, shared=None):
+        build(opt)
+        super().__init__(opt, shared)
+
+        suffix = self.datatype
+        if suffix.startswith('train'):
+            suffix = 'train'
+        elif suffix.startswith('test'):
+            suffix = 'test'
+        elif suffix.startswith('valid'):
+            suffix = 'dev'
+
+        datapath = os.path.join(
+            opt['datapath'],
+            'dialogue_nli',
+            'dnli',
+            'dialogue_nli',
+            'dialogue_nli_' + suffix + '.jsonl',
+        )
+        self._setup_data(datapath)
+
+        self.id = 'dnli'
+        self.reset()
+
+    def _setup_data(self, path):
+        with open(path) as data_file:
+            if 'extra' in path and 'train' in path:
+                line = data_file.readline()
+
+                # trim corrupted JSON
+                line = line[: line.rfind("{")]
+                line = line[: line.rfind(",")] + "]"
+
+                self.data = json.loads(line)
+            else:
+                self.data = json.load(data_file)
+
+    def num_examples(self):
+        return len(self.data)
+
+    def num_episodes(self):
+        return self.num_examples()
+
+    def get(self, episode_idx, entry_idx=0):
+        entry = self.data[episode_idx]
+        entry['episode_done'] = True
+        entry['labels'] = [entry['label']]
+        entry['text'] = entry['sentence1'] + '\n' + entry['sentence2']
+        return entry
+
+
+class ExtrasTeacher(DialogueNliTeacher):
+    def __init__(self, opt, shared=None):
+        build(opt)
+        super().__init__(opt, shared)
```

Comment:
since you are subclassing the teacher above, this teacher will setup the data twice. if you abstract the setting of the datapath for `_setup_data` into a `_path` function (e.g. see [this agent](https://github.com/facebookresearch/ParlAI/blob/master/parlai/tasks/personachat/agents.py)) you can avoid this (and also avoid copying much of the code)

Reply:
I didn't see a clean way to avoid the `super()` call, so I moved the overlapping logic into the parent class!